### PR TITLE
Update to JuMP v1.15 nonlinear syntax

### LINKED ISSUE #61: Revisit Registered Functions
We should explore if it is possible to drop the registered functions in `core/function.jl`.

### LINKED ISSUE #156: HiGHS Dependency
HiGHS should be removed from the dependencies of this package.  It is only required for testing.

### PR DESCRIPTION
**Pull request by odow**
Only substantive change is from a registered function for the head loss to something that is now traceable.

x-ref https://github.com/lanl-ansi/PowerModels.jl/pull/858

Closes #61 
Closes #156

### CODE REVIEW

File: src/form/crd.jl [hunk 1]
```diff
@@ -13,8 +13,8 @@ function constraint_pipe_head_loss(
     dhp, dhn = var(wm, n, :dhp_pipe, a), var(wm, n, :dhn_pipe, a)
 
     # Add relaxed constraints for head loss in the positive and negative directions.
-    c_1 = JuMP.@NLconstraint(wm.model, r * head_loss(qp) <= dhp / L)
-    c_2 = JuMP.@NLconstraint(wm.model, r * head_loss(qn) <= dhn / L)
+    c_1 = JuMP.@constraint(wm.model, r * head_loss(wm, qp) <= dhp / L)
```
Comment (odow) [inline src/form/crd.jl:16]: @ccoffrin here's the function tracing in action. We can now explicitly call `head_loss` and have it return an expression that we substitute in.

File: test/runtests.jl [hunk 2]
```diff
@@ -1,3 +1,6 @@
+import Pkg
+Pkg.pkg"add JuMP#od/nlp-expr"
```
Comment (odow) [inline test/runtests.jl:2]:
```suggestion
Pkg.pkg"add JuMP#master"
```

**Comment by odow** [comment]
I don't understand the 1.6 failure. That just seems wrong. Statistics is a stdlib...

File: test/runtests.jl [hunk 3]
```diff
@@ -1,3 +1,6 @@
+import Pkg
+Pkg.pkg"add JuMP#master"
```
Comment (odow) [inline test/runtests.jl:2]:
```suggestion
Pkg.pkg"add JuMP#master Statistics"
```

File: test/runtests.jl [hunk 4]
```diff
@@ -1,3 +1,6 @@
+import Pkg
+Pkg.pkg"add JuMP#master Statistics"
```
Comment (odow) [inline test/runtests.jl:2]: I don't understand why `Statistics` was necessary. Never seen this problem before...

**Comment by odow** [comment]
@ccoffrin this is ready to go on my end

**Review by tasseff — APPROVED** [review]
Looks great! Thanks, @odow! Super excited by the new nonlinear interface.

**Comment by ccoffrin** [comment]
LGTM.  Before tagging the next release we should also make sure to fix https://github.com/lanl-ansi/WaterModels.jl/issues/156 (which is many be covered in this PR as well)

**Comment by odow** [comment]
I just realized one potential issue: do you have projects that integrate PowerModels and WaterModels? If so, if one use the `@NL` syntax and one uses the new syntax, they're going to error.

**Comment by ccoffrin** [comment]
We do have some packages that combine the infrastructure models packages.  These are the ones I can think of off the top of my head,
https://github.com/lanl-ansi/PowerModelsITD.jl
https://github.com/lanl-ansi/GasPowerModels.jl
https://github.com/lanl-ansi/PowerWaterModels.jl

As long long as the version of each package that updates to JuMP v1.15 is a breaking release then the package version requirements should prevent incompatible versions.  However, I do think we should put a concerted effort into updating the whole ecosystem to the new NL interface, so the packages can still interoperate as they progress.

**Pull request merged**